tools/draincluster/drain/drain.go

```diff
@@ -0,0 +1,244 @@
+		return true, nil
+	}
+
+	// create eviction objects for all <crpName, targetCluster>.
+	for crpName := range h.ClusterResourcePlacementResourcesMap {
+		evictionName := fmt.Sprintf(drainEvictionNameFormat, crpName, h.ClusterName)
+
+		if err := removeExistingEviction(ctx, h.HubClient, evictionName); err != nil {
+			return false, fmt.Errorf("failed to remove existing eviction for CRP %s", crpName)
+		}
+
+		err := retry.OnError(retry.DefaultBackoff, func(err error) bool {
+			return k8errors.IsAlreadyExists(err)
+		}, func() error {
+			eviction := placementv1beta1.ClusterResourcePlacementEviction{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: evictionName,
+				},
+				Spec: placementv1beta1.PlacementEvictionSpec{
+					PlacementName: crpName,
+					ClusterName:   h.ClusterName,
+				},
+			}
+			return h.HubClient.Create(ctx, &eviction)
+		})
+
+		if err != nil {
+			return false, fmt.Errorf("failed to create eviction for CRP %s: %w", crpName, err)
+		}
+	}
+
+	// wait until all evictions reach a terminal state.
+	for crpName := range h.ClusterResourcePlacementResourcesMap {
+		err := wait.ExponentialBackoffWithContext(ctx, retry.DefaultBackoff, func(ctx context.Context) (bool, error) {
+			evictionName := fmt.Sprintf(drainEvictionNameFormat, crpName, h.ClusterName)
+			eviction := placementv1beta1.ClusterResourcePlacementEviction{}
+			if err := h.HubClient.Get(ctx, types.NamespacedName{Name: evictionName}, &eviction); err != nil {
+				return false, fmt.Errorf("failed to get eviction %s: %w", evictionName, err)
+			}
+			return evictionutils.IsEvictionInTerminalState(&eviction), nil
+		})
+
+		if err != nil {
+			return false, fmt.Errorf("failed to wait for evictions to reach terminal state: %w", err)
+		}
+	}
+
+	isDrainSuccessful := true
+	for crpName := range h.ClusterResourcePlacementResourcesMap {
+		evictionName := fmt.Sprintf(drainEvictionNameFormat, crpName, h.ClusterName)
+		eviction := placementv1beta1.ClusterResourcePlacementEviction{}
+		if err := h.HubClient.Get(ctx, types.NamespacedName{Name: evictionName}, &eviction); err != nil {
+			return false, fmt.Errorf("failed to get eviction %s: %w", evictionName, err)
+		}
+		validCondition := eviction.GetCondition(string(placementv1beta1.PlacementEvictionConditionTypeValid))
+		if validCondition != nil && validCondition.Status == metav1.ConditionFalse {
+			// check to see if CRP is missing or CRP is being deleted or CRB is missing.
+			if validCondition.Reason == condition.EvictionInvalidMissingCRPMessage ||
+				validCondition.Reason == condition.EvictionInvalidDeletingCRPMessage ||
+				validCondition.Reason == condition.EvictionInvalidMissingCRBMessage {
+				continue
+			}
+		}
+		executedCondition := eviction.GetCondition(string(placementv1beta1.PlacementEvictionConditionTypeExecuted))
+		if executedCondition == nil || executedCondition.Status == metav1.ConditionFalse {
```

Comment:
This part heavily relies on a few of our coding assumptions here (e.g., valid -> executed, valid = false then executed is not set, valid is never unknown, and executed is never unknown).

Reply:
We use https://github.com/Azure/fleet/pull/1060/files#diff-6e3b2712c782f83d3836cdf31f77c57fc50d0aa4573f4ffa3147d9aacbbfd981R16 util to safeguard ourselves

Reply:
Hi Arvind! It's not a blocker per se but I still believe that we should reply on API semantics for validations in this regard.

Reply:
Plus, even with the function you mentioned, unknown Valid condition can still pass the check, so are unknown Executed conditions.

Reply:
Added TODO will handle in a follow up PR